### ISSUE TITLE
Where-Object and Foreach-object do not escape single quotes in property names

### ISSUE THREAD

**Issue by dkaszews** [PowerShell/PowerShell#24493]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

For PSCustomObject's (and likely others) properties containing characters like spaces and quotes, tab completion automatically quotes and escapes them. As example payload, take:

```pwsh
> $employees = "ID,Employee's name`n1,Alice`n2,Bob"
```

While `$employees[0].E<TAB>` will correctly produce `$employees[0].'Employee''s name'`, doubling single quote to escape it, it is not done for `Where-Object` and `ForEach-Object`. 

### Expected behavior

```console
> $Employees | % E<TAB>
$Employees | % 'Employee''s name'
```

### Actual behavior

```console
> $Employees | % E<TAB>
$Employees | % 'Employee's name'
```

### Error details

```console
Pressing enter waits for input due to unclosed quote.
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.5
PSEdition                      Core
GitCommitId                    7.4.5
OS                             Microsoft Windows 10.0.22631
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

**Comment by dkaszews** [comment]
Paging @MartinGC94 who I believe is local expert on tab completion

**Comment by MartinGC94** [comment]
This can be fixed here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L6882 by changing it from `completionText = $"'{memberName}'";` to `completionText = $"'{memberName.Replace("'", "''")}'";` which is how it's done in other places. Though like in those other places it won't work if the member name contains special single quote chars like `’`.

**Comment by dkaszews** [comment]
Wouldn't it make more sense to try and centralise ilthe code, so that commands use the same escaping as `[0].E<YAB>`?Otherwise there will be more mismatches

**Comment by MartinGC94** [comment]
Definitely. What I suggested before was just a quick fix. Ideally someone should do what this PR tried to do: https://github.com/PowerShell/PowerShell/pull/10226 which is to go through all the completers to make the escaping methods work consistently.  
Personally though I don't feel like doing the work because:  
1: I have a bunch of other pending PRs that I don't want to risk merge conflicts with.  
2: It's rarely something that bothers me because member names are typically simple enough that the current escaping works well enough for me.

**Comment by dkaszews** [comment]
Sure thing, refactors are usually done when there is no other pending work.

**Comment by santisq** [comment]
> This can be fixed here: https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs#L6882 by changing it from `completionText = $"'{memberName}'";` to `completionText = $"'{memberName.Replace("'", "''")}'";` which is how it's done in other places. Though like in those other places it won't work if the member name contains special single quote chars like `’`.
> 
> > <img alt="" width="48" src="https://repository-images.githubusercontent.com/49609581/aad0ad80-cdec-11ea-8248-a37bc0571bfd"><img alt="" height="14" src="https://camo.githubusercontent.com/c7475aac04b72ccdbdab6d5f8abf505c824fdb625ef049acf2c773106c631bff/68747470733a2f2f6769746875622e6769746875626173736574732e636f6d2f66617669636f6e732f66617669636f6e2e737667"> GitHub**[PowerShell/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs at master · PowerShell/PowerShell](https://github.com/PowerShell/PowerShell/blob/master/src/System.Management.Automation/engine/CommandCompletion/CompletionCompleters.cs)**PowerShell for every system! Contribute to PowerShell/PowerShell development by creating an account on GitHub.

Wouldn't it make sense to use `CodeGeneration.EscapeSingleQuotedStringContent` instead of `.Replace` ?

**Comment by MartinGC94** [comment]
I didn't know about that method but I don't see any reason why it should be preferred over `.Replace`.

**Comment by santisq** [comment]
Well, main reason is for what you mentioned in the last comment:

> Though like in those other places it won't work if the member name contains special single quote chars like `’`

This method handles that:

```powershell
[System.Management.Automation.Language.CodeGeneration]::EscapeSingleQuotedStringContent("foo’bar")
# Outputs: foo’’bar
```

**Comment by dkaszews** [comment]
Would something like this work?

```csharp
var completion Text = (memberName.IndexOfAny(s_charactersRequiringQuotes) != 0)
    ? $"'{EscapeSingleQuotedStringContent(memberName)'"
    : memberName;
if (isMethod && addMethodParenthesis)
{
    completionText += "(";
}
```

**Comment by santisq** [comment]
> Would something like this work?
> 
> ```cs
> var completion Text = (memberName.IndexOfAny(s_charactersRequiringQuotes) != 0)
>     ? $"'{EscapeSingleQuotedStringContent(memberName)'"
>     : memberName;
> if (isMethod && addMethodParenthesis)
> {
>     completionText += "(";
> }
> ```

I don't think methods can have a `'` (might be mistaken) but also you shouldn't try to add `(` when using the `-MemberName` + `-ArgumentList` parameter set. `ForEach-Object` will correctly bind the method arguments using spaces, i.e.:

```powershell
class Test {
    [string] Method([string] $foo, [string] $bar) {
        return "$foo $bar"
    }
}

[Test]::new() | % Method hello world
```
